### ISSUE TITLE
NotImplementedError in backprop on on dense-sparse matrices

### ISSUE THREAD

**Issue by mino98** [pytorch/pytorch#102162]
### 🐛 Describe the bug

As Pytorch does not (yet?) support broadcasting on sparse matrices, I implemented a simple autograd class. Forward propagation works fine, but backprop fails with a cryptic:

```
Could not run 'aten::as_strided' with arguments from the 'SparseCPU' backend.
```

Here is a snippet of minimal code to reproduce the issue (including full error message):
https://gist.github.com/mino98/38ef8278470ffdd254c50b428cf1bb38

Thanks

### Versions

Collecting environment information...
PyTorch version: 2.0.1+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.25.2
Libc version: glibc-2.31

Python version: 3.10.11 (main, Apr  5 2023, 14:15:10) [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.15.107+-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          2
On-line CPU(s) list:             0,1
Thread(s) per core:              2
Core(s) per socket:              1
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           79
Model name:                      Intel(R) Xeon(R) CPU @ 2.20GHz
Stepping:                        0
CPU MHz:                         2200.216
BogoMIPS:                        4400.43
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       32 KiB
L1i cache:                       32 KiB
L2 cache:                        256 KiB
L3 cache:                        55 MiB
NUMA node0 CPU(s):               0,1
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Vulnerable; SMT Host state unknown
Vulnerability Meltdown:          Vulnerable
Vulnerability Mmio stale data:   Vulnerable
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:        Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Vulnerable
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.22.4
[pip3] torch==2.0.1+cu118
[pip3] torchaudio==2.0.2+cu118
[pip3] torchdata==0.6.1
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.15.2
[pip3] torchvision==0.15.2+cu118
[pip3] triton==2.0.0
[conda] Could not collect

**Comment by mingfeima** [comment]
@pearu is this expected ?

**Comment by pearu** [comment]
The example backward method can be simplified to
```python
    @staticmethod
    def backward(ctx, grad_output):
        sparse_1d, = ctx.saved_tensors
        grad_sum_sparse = grad_output.sum(dim=tuple(range(1, len(ctx.target_shape))))
        assert (grad_sum_sparse._indices() == sparse_1d._indices()).all()
        return grad_sum_sparse, None

```

With the recent main branch, the example fails with:
```python
  File "/home/pearu/git/pytorch/pytorch-to_dense/torch/_tensor.py", line 488, in backward
    torch.autograd.backward(
  File "/home/pearu/git/pytorch/pytorch-to_dense/torch/autograd/__init__.py", line 204, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: expand is unsupported for Sparse tensors
```
where `expand` on the sparse tensor is called from `sum_backward` function.

The `expand` can be avoided when using
```python
        grad_sum_sparse = grad_output.sum(dim=tuple(range(1, len(ctx.target_shape))), keepdim=True)[:, 0, 0, 0]
```
but that will lead to
```python
 File "/home/pearu/test/pytorch/issue102162.py", line 27, in backward
    grad_sum_sparse = grad_output.sum(dim=tuple(range(1, len(ctx.target_shape))), keepdim=True)[:, 0, 0, 0]
NotImplementedError: Could not run 'aten::as_strided' with arguments from the 'SparseCPU' backend.
```
as reported in this issue.

In sum, I think implementing sparse support for `expand` will fix this issue. Notice that semantically this makes much more sense than implementing sparse support for `as_strided`.
The implementation does not have to be general/complete for the example case here where basically only adding extra dimensions to a one-dimensional sparse tensor is required.

**Comment by mino98** [comment]
Thanks @pearu! 
I can try looking into adding sparse support for `expand` (I assume [here](https://github.com/pytorch/pytorch/blob/f79d2b45fbb2bb47df206e770a525ee420194738/aten/src/ATen/native/TensorShape.cpp#L1102-L1117)), can you share any guidance/tips?

**Comment by pearu** [comment]
Yes, the referenced code is an appropriate place to start introducing sparse support for `expand`. However, notice that [expand](https://pytorch.org/docs/master/generated/torch.Tensor.expand.html) must be a view operation (on `values`) which means that the support will be very restrictive and one should first research if such support will fix the current issue. If it does not then consider eliminating the `expand` call in the given code path of sparse tensor cases altogether. HTH.
